Feature/69399 required project attributes which are not activated for all project

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/69399

# What are you trying to accomplish?

- [x] administration -> projects -> project attributes index
  - [x] add "all projects" indicator
- [x] administration -> projects -> project attributes edit
  - [x] show correct caption in "Projects" tab
  - [x] hide "add project" button
- [x] projects -> project attributes index
  - [x] control toggle box via for_all
  - [x] show "all projects" indicator
- [x] administration -> projects -> project attributes -> new (multiple flavors)
  - [x] add is_for_all checkbox
- [x] double check API
  - [x] check projects API
- [x] Project Initiation Request -> show "all projects" indicator
- [ ] ~Project Initiation Request -> ensure only configured attributes show up in the creation wizard~
  - known bug: https://community.openproject.org/wp/69796 - out of scope for this PR.
- [x] migration: set all fields which are currently `is_required=true` to also have `is_for_all=true` 
  - [x] Warn DevOps, usually there are no migrations allowed after feature freeze

## Screenshots

<img width="1140" height="882" alt="image" src="https://github.com/user-attachments/assets/4a03f837-0d2c-4628-81de-11c2d9f8ad98" />

<img width="2280" height="1774" alt="image" src="https://github.com/user-attachments/assets/b64e9062-a06a-4fca-8cb2-230aebb32620" />

<img width="2280" height="1774" alt="image" src="https://github.com/user-attachments/assets/60339b36-162f-4911-a919-ebec4d03f762" />

<img width="2280" height="1592" alt="image" src="https://github.com/user-attachments/assets/e59083e9-05dc-45dd-9287-f97d5f43826e" />


# What approach did you choose and why?

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#69399] Add "for all projects" checkbox to project custom fields

EinLama committed:
[#69399] Activate project custom fields set "for all"

EinLama committed:
[#69399] project custom field section: for_all controls the toggle

EinLama committed:
[#69399] new project service considers both flags

EinLama committed:
[#69399] projects tab "for_all" caption for project custom fields

EinLama committed:
[#69399] Make specs green

EinLama committed:
[#69399] Additional specs

EinLama committed:
[#69399] New project wizard considers "for_all" cfs

EinLama committed:
[#69399] Fix copy spec

EinLama committed:
[#69399] Show "all projects" info on project attributes

EinLama committed:
[#69399] PIR allows disabling "for_all" attributes

EinLama committed:
[#69399] Project#new and #copy forms show correct custom fields

EinLama committed:
[#69399] Better wording

EinLama committed:
[#69399] Allow PIR attributes to be toggled freely

You can enable/disable required fields.

EinLama committed:
[#69399] Additional specs for custom field form

EinLama committed:
[#69399] Project Custom Field specs for is_for_all

EinLama committed:
[#69399] Migration: is_for_all=true for all required project cfs

EinLama committed:
[#69399] Required fields are mandatory in a PIR

EinLama committed:
[#69399] Fix new custom field projects modal

EinLama committed:
[#69399] Add API specs

EinLama commented:
Dear @opf/ops team, heads up! This PR includes a migration that flips a boolean for **all custom field rows with type=ProjectCustomField**. I added a migration spec and update the records in batches. So I hope this will work without any issues 🤞

EDIT: review is still pending. It will not be merged today, soonest on Monday.

ulferts commented on app/components/projects/settings/project_custom_field_sections/custom_field_row_component.rb:
```diff
@@ -51,9 +51,10 @@ def wrapper_uniq_by
+          @project_custom_field.is_for_all? ||
+            @project_custom_field_project_mappings.any? do |mapping|
+              mapping.custom_field_id == @project_custom_field.id
+            end
```
When enabling a custom field to be active in all projects, it is directly added to the ProjectCustomFieldMappings for all existing projects. Because of that, I don't understand why that addition of  `@project_custom_field.is_for_all? ||` is necessary.

EinLama replied:
That's true, this addition is not necessary. I added it so that for the simple case of `is_for_all=true`, we can skip checking the mappings.

But now that I look at it: we only skip a call to `any?`, but have to introduce additional logic to do that. Not a worthy tradeoff. Would be happy to see this gone.

ulferts commented on app/views/custom_fields/_form.html.erb:
```diff
@@ -205,7 +205,13 @@ See COPYRIGHT and LICENSE files for more details.
+    <div class="form--field">
+      <%= f.check_box :is_for_all %>
+      <div class="form--field-instructions">
+        <p><%= t("custom_fields.instructions.is_for_all") %></p>
```
The phrasing is inconsistent. The instructions speak of custom fields whereas the thing is called project attributes otherwise in the form.

<img width="924" height="195" alt="Image" src="https://github.com/user-attachments/assets/1dbaccd5-a816-4962-8e37-2a4725e14255" />

That problem existed before (and I think there is a bug for it) but now would be a good opportunity to fix it.

ulferts commented on app/forms/projects/settings/custom_fields_form.rb:
```diff
@@ -43,8 +43,11 @@ def additional_custom_field_input_arguments
       private
 
       def custom_fields
+        enabled_custom_fields = model.enabled_custom_field_ids.presence || ProjectCustomField.for_all.pluck(:id)
```
Not a big issue (and probably only an academic problem) but doing it like that, only part of this method's database access are prevented on a subsequent call. The call to `ProjectCustomField.for_all.pluck(:id)` will still execute.

But if you use ProjectCustomField.for_all.select(:id), the DB call would not execute until the statement below is triggered. Additionally, only one SQL statement for both lines are executed after the change.

EinLama replied:
Well spotted and good point!

In addition, there's no reason to execute the first line more than one time, since the result for the query is cached anyways. So all of it could be combined:

```ruby
      def custom_fields
        @custom_fields ||= begin
          enabled_custom_fields = model.enabled_custom_field_ids.presence || ProjectCustomField.for_all.select(:id)

          model
            .available_custom_fields
            .where(id: enabled_custom_fields)
            .required
        end
      end
```

ulferts commented on app/services/projects/concerns/new_project_service.rb:
```diff
@@ -104,11 +104,11 @@ def disable_custom_fields_with_empty_values(new_project)
 
       custom_field_ids = project.custom_values
-        .select { |cv| cv.value? || cv.required? }
+        .select { |cv| cv.value? || cv.is_for_all? }
```
There is a bug here that existed before. 

If you have a custom field with a default value, it will return in 
```
project.custom_values.select { |cv| cv.value? }
```

If it is a required field, it will also show up in the third step of the project creation (even if it is not an `is_for_all`). A non required field will not show up there but it will be activated within the new project wrongfully.

We would need to check here if the value was actually provided by the user or if it is is the default value. We should be able to get the params the service was called with and do the check on that.

Does not have to be part of this PR but if it isn't. We should open a bug

ulferts commented on app/forms/custom_fields/calculated_values/details_form.rb:
```diff
@@ -65,7 +65,13 @@ class DetailsForm < ApplicationForm
           name: :is_required,
           label: I18n.t("activerecord.attributes.project_custom_field.is_required"),
-          caption: I18n.t("custom_fields.instructions.is_required_for_project")
+          caption: I18n.t("custom_fields.instructions.is_required")
```
I think I would remove the "required" field for calculated value. That value cannot be filled in by the user directly. So even if it is "required", it is not. That would also mean that in the migration, calculated values' "is_required" field should be unset.

EinLama replied:
Huh, that's a good point 🤔

Then, the migration would require a `down` method that sets `is_required=true` again for calculated values in case of a rollback.

ulferts commented on app/views/custom_fields/_form.html.erb:
```diff
@@ -205,7 +205,13 @@ See COPYRIGHT and LICENSE files for more details.
       <%= f.check_box :is_required %>
       <div class="form--field-instructions">
-        <p><%= t("custom_fields.instructions.is_required_for_project") %></p>
+        <p><%= t("custom_fields.instructions.is_required") %></p>
```
Same as with calculated values I don't see the reason for allowing boolean custom fields to be marked as required. To the user, the fact that the value is not filled out has the same meaning as having unset the value. In the eyes of the user, both would count as "the field has the correct (false) value". 

The problem with this additional requirement is that the other customizables don't treat booleans in that fashion. 

It might therefore be better to change it consistently for all customizables in a separate PR.

ulferts replied:
The current representation of a required boolean cf on the project copy form supports the idea:

<img width="333" height="50" alt="Image" src="https://github.com/user-attachments/assets/eef9f912-09b9-4948-85f8-597095e68760" />

There is no asterisks on the label other than for the other required fields.

But on the other hand, using the "require" flag leads to the field showing up early so users can think about it.

ulferts commented on db/migrate/20251211160744_set_is_for_all_for_required_project_custom_fields.rb:
```diff
@@ -0,0 +1,46 @@
+    required_pcfs = ProjectCustomField.where(is_required: true)
+    in_configurable_batches(required_pcfs) do |batch|
+      batch.update_all(is_for_all: true)
+    end
```
`update_all` [will not instantiate the record but rather write to the database directly.](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-update_all) Because of that, doing it in batches is unnecessary.

You can simply do

```
ProjectCustomField.where(is_required: true).update_all(is_for_all: true)
```

But depending on how we fare with the other comments (e.g. whether calculated values should possibly loose their "is_required" state) it might get more complicated again.

EinLama replied:
I added the batching since I wasn't sure how an `UPDATE` statement would perform on very large tables. So less of a Ruby safeguard and more of a database-lifebelt 🛟 Not sure if needed, but if it doesn't hurt, it's better to be safe than sorry.

ulferts reviewed (changes requested):
For a change of this character, I am surprised that I did not find more. But I found a couple of things. Only some are actual bugs in the code. For the most part, addressing the comment will require a discussion about requirements, I fear.

There is also one behaviour that I couldn't attribute to a code change. It is related to the API.

If you have multiple required cfs configured that are also active for all projects and don't have default values, then sending a 

```
http://localhost:3000/api/v3/projects
```

With the body
```
{
    "name": "A project created via the api"
}
```

will yield the expected error messages of values not being set for the custom fields.
```
{
    "_type": "Error",
    "errorIdentifier": "urn:openproject-org:api:v3:errors:MultipleErrors",
    "message": "Multiple field constraints have been violated.",
    "_embedded": {
        "errors": [
            {
                "_type": "Error",
                "errorIdentifier": "urn:openproject-org:api:v3:errors:PropertyConstraintViolation",
                "message": "Admin only cf (required) can't be blank.",
                "_embedded": {
                    "details": {
                        "attribute": "customField676"
                    }
                }
            },
            {
                "_type": "Error",
                "errorIdentifier": "urn:openproject-org:api:v3:errors:PropertyConstraintViolation",
                "message": "Required List cf (for all) can't be blank.",
                "_embedded": {
                    "details": {
                        "attribute": "customField514"
                    }
                }
            },
            {
                "_type": "Error",
                "errorIdentifier": "urn:openproject-org:api:v3:errors:PropertyConstraintViolation",
                "message": "Required date cf (not for all) can't be blank.",
                "_embedded": {
                    "details": {
                        "attribute": "customField677"
                    }
                }
            }
        ]
    }
}
```

But when then sending the body of
```
{
    "customField525": 5,
    "name": "A project created via the api"
}
```
 the project is created successfully. Please notice that the custom field used here is not even one of the required ones. It seems that when sending any custom field value, the others are ignored.

ulferts committed:
remove required from calculated and bool

ulferts committed:
remove ability of boolean cfs to be required

ulferts committed:
migrate is_required for bool and calculated values

ulferts committed:
consolidate custom field form and component

ulferts committed:
hide is_filter field for hierarchy and weighted list as well

The field had no effect. Regardless of whether it was true or not, a filter existed in the project list

ulferts committed:
remove unnecessary dublicate condition

ulferts committed:
differentiate between project attribute and custom field in captions

ulferts committed:
inline sql statement into another one to avoid one request

ulferts committed:
move the check for component into the component

ulferts reviewed (approved):
Two bug reports have been opened for the topics that were not addressed within this PR.
